bin/strata-client/src/args.rs

```diff
@@ -1,171 +1,230 @@
+}
 
-    /// Max retries for Bitcoin RPC calls.
-    #[argh(option, description = "max retries for bitcoin RPC (default: 3)")]
-    pub bitcoind_retry_count: Option<u8>,
+impl Args {
+    /// Overrides config. First overrides with the generic overrides passed via `-o` and then
+    /// overrides the result with some of the commonly passed args. Returns if config was overridden
+    /// or not.
+    pub fn override_config(&self, config: &mut Config) -> anyhow::Result<bool> {
+        // Override using -o params.
+        let mut overridden = self.override_generic(config)?;
+
+        // Override by explicitly parsed args like datadir, rpc_host and rpc_port.
+        if let Some(datadir) = &self.datadir {
+            config.client.datadir = datadir.into();
+            overridden = true
+        }
+        if let Some(rpc_host) = &self.rpc_host {
+            config.client.rpc_host = rpc_host.to_string();
+            overridden = true
+        }
+        if let Some(rpc_port) = &self.rpc_port {
+            config.client.rpc_port = *rpc_port;
+            overridden = true
+        }
+        Ok(overridden)
+    }
 
-    /// Timeout duration for btc request retries in ms. Defaults to `1000`.
-    #[argh(
-        option,
-        description = "max interval between bitcoin RPC retries in ms (default: 1000)"
-    )]
-    pub bitcoind_retry_interval: Option<u64>,
+    /// Override config using the generic overrides. The idea here is to first convert the
+    /// [`Config`] to json and then apply the overrides by splitting the key by a dot.
+    fn override_generic(&self, config: &mut Config) -> anyhow::Result<bool> {
+        let original = config.clone();
+        // Convert config as json
+        let mut json_config = to_value(&mut *config).context("Config json serialization failed")?;
 
-    #[argh(option, short = 'n', description = "L1 network to run on")]
-    pub network: Option<Network>,
+        for (path, val) in parse_overrides(&self.overrides)?.iter() {
+            apply_override(path, val, &mut json_config)?;
+        }
+        // Convert back to json.
+        *config = from_value(json_config)
+            .context("Should be able to create Config from serde json Value")?;
+        Ok(original != *config)
+    }
+}
 
-    #[argh(switch, description = "start sequencer bookkeeping tasks")]
-    pub sequencer: bool,
+type Override = (Vec<String>, String);
+
+/// Parse valid overrides. This first splits the entries by '=' to get key and value and then splits
+/// the key by '.' which is the update path.
+fn parse_overrides(overrides: &[String]) -> anyhow::Result<Vec<Override>> {
+    let mut result = Vec::new();
+    for item in overrides {
+        let (key, value) = item
+            .split_once("=")
+            .ok_or(anyhow!("Invalid override: must be in 'key=value' format"))?;
+        let path: Vec<_> = key.split(".").map(|x| x.to_string()).collect();
+        result.push((path, value.to_string()));
+    }
+    Ok(result)
+}
```

Comment:
Functional programming heuristic: The way this is written, it's not possible to parse just a single override string.  You always have to parse a *list of* override strings.  The body of the loop is where the hard part is happening, and then we don't need a function just to transform elements of a list.

Reply:
Truee.